### PR TITLE
Replace ink_atomic with std::atomic

### PR DESCRIPTION
**Pull request by PSUdaemon**
We should remove ink_atomic and replace it with std::atomic.

### CODE REVIEW

**Comment by dragon512** [comment]
Oh no! build failure on Linux!!!

**Comment by dragon512** [comment]
We might need to add 
`-std=c11`

This headers should exist for gcc 4.9 or better

**Comment by SolidWallOfCode** [comment]
If it's just plugins written in C that are the problem, I say put the old `ink_atomic...` in a separate header file named `c99_atomics.h` or some such and use it only in C plugins.

**Comment by bryancall** [comment]
[approve ci autest]

File: plugins/s3_auth/s3_auth.cc [hunk 1]
```diff
@@ -212,14 +213,14 @@ class S3Config
...
+    _ref_count++;
   }
 
   void
   release()
   {
-    TSDebug(PLUGIN_NAME, "ref_count is %d", _ref_count);
-    if (1 >= ink_atomic_decrement(&_ref_count, 1)) {
+    TSDebug(PLUGIN_NAME, "ref_count is %d", (int)_ref_count);
+    if (1 >= _ref_count--) {
```
Comment (bryancall) [inline plugins/s3_auth/s3_auth.cc]: Don't you want --_ref_count ?

Comment (bryancall) [inline plugins/s3_auth/s3_auth.cc]: Thought this might be a problems, but it isn't.

**Comment by bryancall** [comment]
[approve ci autest]

File: plugins/s3_auth/s3_auth.cc (hunk 1 above)
Comment (PSUdaemon) [reply plugins/s3_auth/s3_auth.cc]: Which part? I am changing that to a C++ static_cast for @SolidWallOfCode.

Comment (bryancall) [reply plugins/s3_auth/s3_auth.cc]: I was talking about the postfix decrement and when it happens.  It happens before the conditional.

**Comment by PSUdaemon** [comment]
@jpeach if I go down that path for the inc/dec should I also do it for load/store? Basically every use of a std::atomic will require some modification. To me this kinda defeats the purpose of std::atomic. Just calling a method instead of wrapping it in a function.

**Comment by PSUdaemon** [comment]
I know there are a ton of failures right now, but I wanted @jpeach to look at what I meant with the load/store. [Here is an example.](https://github.com/apache/trafficserver/pull/2633/commits/a717f0d8c9ee98ba137a78e5db531425ae2c9c75#diff-da93c916d77790de361cf0cfc9c854d5)

**Comment by jpeach** [comment]
@PSUdaemon I think that atomic ops deserve to be made very explicit in the code, even at the cost of extra typing ;)

**Comment by bryancall** [comment]
This PR hasn't been updated in more then 2 months and I am closing out PRs that haven't had any updates.  Please feel free to reopen if want to update it.

**Pull request closed without merging**